server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -257,6 +268,44 @@ public ClusterState execute(ClusterState currentState) {
                 // Store newSnapshot here to be processed in clusterStateProcessed
                 List<String> indices = Arrays.asList(indexNameExpressionResolver.concreteIndexNames(currentState, request));
 
+                List<SnapshotFeatureInfo> featureStates = Collections.emptyList();
```

Comment:
I think we need BwC here. We shouldn't be triggering ta snapshot that includes `featureStates` as long as there's old master nodes in the cluster that don't understand `featureStates` yet. Even if it might work out magically now (because we redundantly store what to snapshot in `indices` it's a potential point of failure where we move to a different state machine on master failover.

Reply:
To clarify, what you're recommending here is not using `featureStates` in a mixed cluster where some master nodes are <`FEATURE_STATES_VERSION`?

In that case, f there's a request that explicitly uses feature states, we can just throw an exception that that feature isn't allowed in mixed clusters. But if we have a create snapshot request that implicitly uses feature states (`include_global_state: true`), should we use the pre-7.12.0 behavior regardless of the current version of the master? I don't like silently using the previous behavior like that, but it seems like we'd still want to be able to take a snapshot in that case.

Reply:
>  the pre-7.12.0 behavior regardless of the current version of the master?

I think that's the correct approach as long as there's old version nodes in the cluster. At least, it's what we currently do for all the snapshot features/changes, we always keep using the old state machine as long as there's nodes of an old version around. It's just the by far easiest BwC behavior to reason about.

> In that case, f there's a request that explicitly uses feature states, we can just throw an exception that that feature isn't allowed in mixed clusters

Yea I'd throw here if the request explicitly requests feature states. If it doesn't specify anything, just using the previous behavior seems the way to go.